crates/e2e-test-utils/src/testsuite/mod.rs

```diff
@@ -1,42 +1,214 @@
     env: Environment<I>,
+    /// Store action results for later assertions
+    results: ResultStore,
 }
 
-impl<I> Runner<I> {
+impl<I: 'static> Runner<I> {
+    /// Create a new test runner with the given environment
+    pub fn new(instance: I) -> Self {
+        Self { env: Environment { instance, ctx: () }, results: ResultStore::default() }
+    }
 
-    /// Execute an action
-    pub async fn execute<A>(&self, action: A) -> eyre::Result<()>
-    where A: Action<I>
-    {
+    /// Execute an action and store its result
+    pub async fn execute(&mut self, action: ActionBox<I>) -> Result<()> {
+        let result = action.execute(&self.env).await?;
+
+        match result {
+            ActionResult::Payload(id, payload) => {
+                self.results.store_payload(id, payload);
+                Ok(())
+            }
+            ActionResult::BlockHash(id, hash) => {
+                self.results.store_block_hash(id, hash);
+                Ok(())
+            }
+            ActionResult::TransactionHash(id, hash) => {
+                self.results.store_transaction_hash(id, hash);
+                Ok(())
+            }
+            ActionResult::Value(id, value) => {
+                self.results.store_value(id, value);
+                Ok(())
+            }
+            ActionResult::Bool(id, value) => {
+                self.results.store_bool(id, value);
+                Ok(())
+            }
+            ActionResult::None => Ok(()),
+        }
+    }
+
+    /// Run an assertion against the stored results
+    pub async fn assert(&self, assertion: AssertionBox) -> Result<()> {
+        assertion.assert(&self.results).await
+    }
 
-        todo!()
+    /// Execute a sequence of actions
+    pub async fn run_actions(&mut self, actions: Vec<ActionBox<I>>) -> Result<()> {
+        for action in actions {
+            self.execute(action).await?;
+        }
+        Ok(())
     }
 
+    /// Run a complete test scenario with setup, actions, and assertions
+    pub async fn run_scenario(
+        &mut self,
+        setup: Option<Setup>,
+        actions: Vec<ActionBox<I>>,
+        assertions: Vec<AssertionBox>,
+    ) -> Result<()> {
+        if let Some(setup) = setup {
+            setup.apply(&mut self.env).await?;
+        }
+
+        for action in actions {
+            self.execute(action).await?;
+        }
+
+        for assertion in assertions {
+            self.assert(assertion).await?;
+        }
+
+        Ok(())
+    }
 }
 
 /// Represents a test environment.
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 pub struct Environment<I> {
-   /// The instance against which we can run tests.
-   pub instance: I,
-   pub ctx: (),
+    /// The instance against which we can run tests.
+    pub instance: I,
+    /// Context.
+    pub ctx: (),
+}
+
+/// A store for action results that can be used in assertions
+#[derive(Default, Debug)]
+pub struct ResultStore {
+    payloads: HashMap<String, Arc<ExecutionPayload>>,
```

Comment:
we can't use concrete types like this because those are network specific

Reply:
yep, no more `ResultStore` now that actions and assertions are merged